packages/tailwindcss/src/index.ts

```diff
@@ -153,17 +185,48 @@ export function compile(css: string): {
 
         // Replace the `@apply` rule with the actual utility classes
         {
+          let newNodes: AstNode[] = []
+
+          // Collect all user-defined classes for the current candidates that we
+          // need to apply.
+          for (let candidate of candidates) {
+            let nodes = userDefinedApplyables.get(candidate)
+            if (!nodes) continue
+
+            for (let child of nodes) {
+              newNodes.push(structuredClone(child))
+            }
+          }
+
           // Parse the candidates to an AST that we can replace the `@apply` rule with.
           let candidateAst = compileCandidates(candidates, designSystem, {
             onInvalidCandidate: (candidate) => {
+              // When a candidate is invalid, we want to first verify that the
+              // candidate is a user-defined class or not. If it is, then we can
+              // safely ignore this. If it's not, then we throw an error because
+              // the candidate is unknown.
+              //
+              // The reason we even have to check user-defined classes is
+              // because it could be that the user defined CSS like that is also
+              // a known utility class. For example, the following CSS would be:
+              //
+              // ```css
+              // .flex {
+              //   --display-mode: flex;
+              // }
+              // ```
+              //
+              // When the user then uses `@apply flex`, we want to both apply
+              // the user-defined class and the utility class.
```

Comment:
```suggestion
              // We must pass in user-defined classes and then filter them out
              // here because, while they are usually not known utilities, the
              // user can define a class that happens to *also* be a known
              // utility. For example, given the following, `flex` counts as
              // both a user-defined class and a known utility:
              //
              // ```css
              // .flex {
              //   --display-mode: flex;
              // }
              // ```
              //
              // When the user then uses `@apply flex`, we want to both apply
              // the user-defined class and the utility class.
```